### PR TITLE
Transparency

### PR DESCRIPTION
**Pull request by nifker**
This will partially resolve https://github.com/emoon/rust_minifb/issues/133
Implemented:
- X11
- Wayland
- Redox(couldnt test so far)

### CODE REVIEW

File: examples/transparent.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+use minifb::{Key, ScaleMode, Window, WindowOptions};
+
+fn main() {
+    // Allocate the output buffer.
+    let buf = vec![0x00AAFF33; 320*480];
+
+    let mut window = Window::new(
+        "Press ESC to exit",
+        320 as usize,
```
Comment (emoon) [inline examples/transparent.rs:11]: shouldn't need to say `value as usize` here

**Comment by nifker** [comment]
@cole-h Are you interested to help here?

File: src/lib.rs [hunk 2]
```diff
@@ -187,6 +187,9 @@ pub struct WindowOptions {
     pub scale_mode: ScaleMode,
     /// Should the window be the topmost window (default: false)
     pub topmost: bool,
+    /// Specifies whether or not the window is allowed to draw transparent pixels(default: false)
```
Comment (cole-h) [inline src/lib.rs:190]:
```suggestion
    /// Specifies whether or not the window is allowed to draw transparent pixels (default: false)
```

File: src/os/unix/x11.rs [hunk 3]
```diff
@@ -90,8 +90,8 @@ struct DisplayInfo {
 }
 
 impl DisplayInfo {
-    fn new() -> Result<DisplayInfo> {
-        let mut display = Self::setup()?;
+    fn new(transparency: bool) -> Result<DisplayInfo> {
```
Comment (cole-h) [inline src/os/unix/x11.rs:93]: I don't really like that this takes `transparency` as an argument, but I don't have any suggestions to improve this...

File: src/os/unix/x11.rs [hunk 4]
```diff
@@ -115,10 +115,30 @@ impl DisplayInfo {
...
+                (lib.XMatchVisualInfo)(
+                    display,
+                    (lib.XDefaultScreen)(display),
+                    32,
+                    xlib::TrueColor,
+                    &mut vinfo as *mut _,
+                );
+                depth = vinfo.depth;
+                visual = vinfo.visual;
+                screen = vinfo.screen;
```
Comment (cole-h) [inline src/os/unix/x11.rs:134]: Just a style nit, but would be nice to follow declaration order above and assignment order below:
```suggestion
                screen = vinfo.screen;
                visual = vinfo.visual;
                depth = vinfo.depth;
```

File: examples/transparent.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+extern crate minifb;
```
Comment (cole-h) [inline examples/transparent.rs:1]: I thought Rust 2018 made this unnecessary.

File: src/os/unix/x11.rs (hunk 3 above)
Comment (nifker) [reply src/os/unix/x11.rs:93]: The x11 one currently doesnt even work, when you try the transparency example: it produces a BadMatch on XPutImage.

Comment (nifker) [reply src/os/unix/x11.rs:93]: Im currently trying to find out where the problem is :/

**Comment by emoon** [comment]
Question:
With transparency I assume you can set how visible a window should be? Maybe it would be better with an `Option<f32>` instead of a bool where the `Some(0.5)` would be 50% transparency, `Some(0.75)` 75% etc

**Comment by nifker** [comment]
@emoon This would be possible to implement in theory, but needs to be simulated. The plan here is to allow alpha transparency for windows, where the dev can draw (semi)transparent RGBA data to the window.

**Comment by emoon** [comment]
Alright. Then I guess a bool makes more sense

**Comment by nifker** [comment]
X11 is now working too(but requires an compositor).
@emoon Can you research on how to do this on OSX and implement the code for it?

File: src/os/unix/x11.rs (hunk 3 above)
Comment (nifker) [reply src/os/unix/x11.rs:93]: I actually fixed it, open to improvements now.

**Comment by nifker** [comment]
Ok the windows one will take longer, because windows requires pre-multiplied colours https://github.com/godotengine/godot/pull/14622/commits/ddae098deec89d67228a74623ebe7f5db39f425b#diff-33050a8c8ed21f4806d591c06f251d48R1720-R1740

**Comment by nifker** [comment]
Seems like I am not able to currently implement transparency so far(as of time and experience with windows and osx).
@emoon Would it be possible to review this and merge the linux/bsd/etc. implementation and the windows/osx implementation would come later?

**Comment by emoon** [comment]
Yes that is fine. Should I merge these changes now?

**Comment by nifker** [comment]
@emoon Next time I push Id be ready.

File: src/lib.rs [hunk 6]
```diff
@@ -220,6 +224,12 @@ impl Window {
     ///  .expect("Unable to open Window");
     /// ```
     pub fn new(name: &str, width: usize, height: usize, opts: WindowOptions) -> Result<Window> {
+        if opts.transparency {
+            assert!(
```
Comment (nifker) [inline src/lib.rs:228]: @emoon Should we assert here or return an Err value?

Comment (emoon) [reply src/lib.rs:228]: Return an error code. `assert!` will only be present in debug config.

Comment (nifker) [reply src/lib.rs]: Ok Im done here

**Comment by emoon** [comment]
ok, good to merge now?

**Comment by nifker** [comment]
Yes

**Pull request merged**